File: src/main/java/com/dumbdogdiner/warrior/WarriorPlaceholders.java
package com.dumbdogdiner.warrior;

import com.dumbdogdiner.warrior.api.arena.Arena;
import com.dumbdogdiner.warrior.api.translation.TimeUtil;
import com.dumbdogdiner.warrior.api.user.WarriorUser;
import com.dumbdogdiner.warrior.api.sessions.ArenaSession;
import com.dumbdogdiner.warrior.api.sessions.Session;
import com.dumbdogdiner.warrior.managers.ArenaManager;
import com.dumbdogdiner.warrior.managers.PlayerManager;
import com.dumbdogdiner.warrior.utils.TranslationUtil;
import me.clip.placeholderapi.expansion.PlaceholderExpansion;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.NotNull;

public class WarriorPlaceholders extends PlaceholderExpansion {

    @Override
    public boolean canRegister() {
        return true;
    }

    @Override
    public @NotNull String getIdentifier() {
        return "warrior";
    }

    @Override
    public @NotNull String getAuthor() {
        return Warrior.getInstance().getDescription().getAuthors().get(0);
    }

    @Override
    public @NotNull String getVersion() {
        return Warrior.getInstance().getDescription().getVersion();
    }

    @Override
    public String onPlaceholderRequest(Player p, String identifier) {
        String[] args = identifier.split("_");
        if(args.length < 1) return identifier;
        if(args[0].equalsIgnoreCase("player")) {
            WarriorUser user = PlayerManager.get(p.getUniqueId());
            if(user == null) return identifier;
            switch(args[1]) {
                case "name":
                    return user.getName();
                case "timeplayed":
                    long time = user.getTotalTime() + (System.currentTimeMillis() - user.getLastJoin());
                    return TimeUtil.formatDuration(time);
                case "firstjoin":
                    return TimeUtil.formatDate(user.getSettings().getLanguage(), user.getFirstJoin());
                case "id":
                case "userid":
                case "uuid":
                    return user.getUserId().toString();
                case "ping":
                    return user.getPing() + "ms";
                case "title":
                    return user.getSettings().getTitle().getTitle();
                case "deathsound":
                    return user.getGameplaySettings().getActiveSound().friendlyName();
                case "kills":
                    return Integer.toString(user.getKills());
                case "deaths":
                    return Integer.toString(user.getDeaths());
                case "kdr":
                    return Double.toString(user.getKDR());
                case "coins":
                    return Integer.toString(user.getCoins());
                case "arena":
                    Session session = user.getSession();
                    if(!(session instanceof ArenaSession)) return "(none)";
                    return ((ArenaSession)session).getArena().getName();
            }
        } else if(args[0].equalsIgnoreCase("arena")) {
            Arena arena = null;
            String argument = "";

            WarriorUser user = PlayerManager.get(p.getUniqueId());
            if(user == null) return identifier;

            // if no argument provided, get the arena at the players location
            if(args.length == 2) {
                for(Arena a : ArenaManager.getArenas()) {
                    if(a.getBounds().contains(user.getLocation())) {
                        arena = a;
                        break;
                    }
                }

                argument = args[1];
            } else if(args.length == 3) {
                if(ArenaManager.get(args[1]) == null) return identifier;
                arena = ArenaManager.get(args[1]);

                argument = args[2];
            }

            if(arena == null) return identifier;

            switch (argument) {
                case "name":
                    return arena.getName();
                case "id":
                    return String.valueOf(arena.getId());
                case "flags":
                    return arena.getFlags().toString();
                case "rating":
                    return String.valueOf(arena.getMetadata().averageRating());
                case "creator":
                    return arena.getMetadata().getCreator();
                case "timestamp":
                    return TimeUtil.formatDate(user.getSettings().getLanguage(), arena.getMetadata().getCreatedAt());
                case "description":
                    return String.join("\n", arena.getMetadata().getDesc());
            }
        }

        return identifier;
    }
}

File: warrior/src/main/kotlin/com/dumbdogdiner/warrior/WarriorPlaceholders.kt
package com.dumbdogdiner.warrior

import me.clip.placeholderapi.expansion.PlaceholderExpansion
import com.dumbdogdiner.warrior.api.sessions.ArenaSession
import com.dumbdogdiner.warrior.api.arena.Arena
import com.dumbdogdiner.warrior.api.translation.TimeUtil
import com.dumbdogdiner.warrior.user.User
import org.bukkit.entity.Player

class WarriorPlaceholders : PlaceholderExpansion() {

    // TODO: This class is untested!

    override fun canRegister(): Boolean {
        return true
    }

    override fun getIdentifier(): String {
        return "warrior"
    }

    override fun getAuthor(): String {
        return WarriorPlugin.instance.description.authors[0]
    }

    override fun getVersion(): String {
        return WarriorPlugin.instance.description.version
    }

    override fun onPlaceholderRequest(p: Player, identifier: String): String {
        val args = identifier.split("_").toTypedArray()

        if (args.isEmpty()) return identifier
        if (args[0].equals("player", ignoreCase = true)) {
            val user: User = WarriorPlugin.instance.userCache[p.uniqueId] ?: return identifier
            when (args[1]) {
                "name" -> return user.name
                "timeplayed" -> {
                    val time = user.totalTime + (System.currentTimeMillis() - user.lastJoin)
                    return TimeUtil.formatDuration(time)
                }
                "firstjoin" -> return TimeUtil.formatDate(user.settings.language, user.firstJoin)
                "id", "userid", "uuid" -> return user.userId.toString()
                "ping" -> return user.ping.toString() + "ms"
                "title" -> return user.settings.title.title
                "deathsound" -> return user.gameplaySettings.activeSound.friendlyName()
                "kills" -> return user.kills.toString()
                "deaths" -> return user.deaths.toString()
                "kdr" -> return user.kdr.toString()
                "coins" -> return user.coins.toString()
                "arena" -> {
                    val session = user.session
                    return if (session !is ArenaSession) "(none)" else session.arena.name
                }
            }
        } else if (args[0].equals("arena", ignoreCase = true)) {
            var arena: Arena? = null
            var argument = ""
            val user: User = WarriorPlugin.instance.userCache[p.uniqueId]
                ?: return identifier

            // if no argument provided, get the arena at the players location
            if (args.size == 2) {
                for (arenaIt in WarriorPlugin.instance.arenaManager.arenas) {
                    if (arenaIt.bounds.contains(user.location)) {
                        arena = arenaIt
                        break
                    }
                }
                argument = args[1]
            } else if (args.size == 3) {
                arena = WarriorPlugin.instance.arenaManager[args[1]] ?: return identifier
                argument = args[2]
            }
            if (arena == null) return identifier
            when (argument) {
                "name" -> return arena.name
                "id" -> return arena.id.toString()
                "flags" -> return arena.flags.toString()
                "rating" -> return java.lang.String.valueOf(arena.metadata.averageRating())
                "creator" -> return arena.metadata.creator
                "timestamp" -> return TimeUtil.formatDate(user.settings.language, arena.metadata.createdAt)
                "description" -> return arena.metadata.desc.joinToString(separator = "\n")
            }
        }
        return identifier
    }
}